packages/analytics-browser/src/attribution/campaign-tracker.ts

```diff
@@ -80,33 +88,38 @@ export class CampaignTracker implements ICampaignTracker {
       return identify.unset(key);
     }, new Identify());
 
-    const pageViewEvent: BaseEvent = {
-      event_type: 'Page View',
-      event_properties: {
-        page_title: /* istanbul ignore next */ (typeof document !== 'undefined' && document.title) || '',
-        page_location: /* istanbul ignore next */ (typeof location !== 'undefined' && location.href) || '',
-        page_path: /* istanbul ignore next */ (typeof location !== 'undefined' && location.pathname) || '',
-      },
-    };
-    return {
-      ...createIdentifyEvent(identifyEvent),
-      ...(this.trackPageViews && pageViewEvent),
-    };
+    return createIdentifyEvent(identifyEvent);
   }
 
-  async send(isNewSession: boolean) {
-    if (this.disabled) {
-      return;
-    }
-    const currentCampaign = await this.parser.parse();
-    const previousCampaign = await this.getCampaignFromStorage();
-    if (!isNewSession) {
-      if (!this.trackNewCampaigns || !this.isNewCampaign(currentCampaign, previousCampaign)) {
-        return;
+  trackOn(filter: PageTrackingFilter = undefined, callback: (currentCampaing: Campaign) => unknown) {
+    const { isNewCampaign, currentCampaign } = this.getCurrentState();
+
+    switch (filter) {
+      case 'onAttribution': {
+        return isNewCampaign && callback(currentCampaign);
       }
-      this.onNewCampaign(currentCampaign);
+      default: {
+        if (typeof filter === 'function') {
+          return filter() && callback(currentCampaign);
+        }
+        return callback(currentCampaign);
+      }
+    }
+  }
+
+  async onStateChange(callback: (state: { isNewCampaign: boolean; currentCampaign: Campaign }) => Promise<unknown>) {
```

Comment:
What is onStateChange meant to do? Neither of the places its used use the currentState passed back to the callback. It's confusing that unrelated components are triggering saveCampaignToStorage.

Reply:
Initially I added "onNewSession" event listener here to trigger this callback when there is a new session. Also we can support single page app in this function, so if a page is changed, we can trigger this callback to send out identity event or page view event. 

Both attribution plugin and pageTracking plugin link to this function. Each time the page state changes (session, campaign, or page view for SPA), we can run the MA strategy again.

Reply:
Ok that makes sense, but onNewSession is a bit of a misnomer since the passed callback is always called, even if there's no new session. Maybe `onPageChange`?